lenskit-core/src/main/java/org/lenskit/util/math/Vectors.java

```diff
@@ -326,6 +328,93 @@ public static RealVector matrixRow(RealMatrix mat, int row) {
         return new RowView(mat, row);
     }
 
+    /**
+     * Transpose of a matrix
+     * Convert a row view of rating matrix to a column view and vice versa
+     * After conversion, the key of empty row/column in the input matrix will be eliminated in the transpose matrix.
+     * So calling transposeMap(transposeMap(X)).keySet() may not equal to X.keySet() unless X has no mapping of key to empty Long2Double map.
+     *
+     * @param matrix The matrix
+     * @return The transpose of the matrix
+     */
+    public static Long2ObjectMap<Long2DoubleMap> transposeMap(Long2ObjectMap<Long2DoubleMap> matrix) {
+        Long2ObjectMap<Long2DoubleMap> mapT = new Long2ObjectOpenHashMap<>();
+        for (Map.Entry<Long, Long2DoubleMap> rowEntry : matrix.entrySet()) {
+            long rowNum = rowEntry.getKey();
+            Long2DoubleMap rowValue = rowEntry.getValue();
+            for (Map.Entry<Long, Double> entry : rowValue.entrySet()) {
+                long colNum = entry.getKey();
+                double Value = entry.getValue();
+                Long2DoubleMap rowOfMapT = mapT.get(colNum);
+                if (rowOfMapT == null) rowOfMapT = new Long2DoubleOpenHashMap();
+                rowOfMapT.put(rowNum, Value);
+                mapT.put(colNum, rowOfMapT);
+            }
+        }
+        return mapT;
+    }
+
+    /**
+     * Compute the addition of two vectors
+     * @param a The first vector
+     * @param b The second vector
+     * @return A new vector of the addition of two vectors {@code a} + {@code b}
+     */
+    public static Long2DoubleMap combineAndLimit(Long2DoubleMap a, Long2DoubleMap b) {
```

Comment:
I think this one should be called `add`, and the old `combine` should be `combineAndLimit`.